### PR TITLE
core/types, rlp: optimize derivesha

### PR DESCRIPTION
**Pull request by holiman**
This PR contains a minor optimization in derivesha, by exposing the rlp int-encoding and making use of it to write integers directly to a buffer (an rlp integer is known to never require more than `9` bytes total). 
This idea (using `rlp.PutInt`) might be useful in other places too. Is it a good idea? Consider it a suggestion.  

The code assumes, just as before, that the `hasher` (a trie) will copy the key internally, which it does when doing `keybytesToHex(key)`. 

```
name                       old time/op    new time/op    delta
DeriveSha200/std_trie-6       590µs ± 8%     546µs ± 4%   -7.43%  (p=0.000 n=9+9)
DeriveSha200/stack_trie-6     492µs ± 2%     458µs ± 5%   -7.05%  (p=0.000 n=9+9)

name                       old alloc/op   new alloc/op   delta
DeriveSha200/std_trie-6       275kB ± 0%     274kB ± 0%   -0.61%  (p=0.000 n=10+10)
DeriveSha200/stack_trie-6    55.0kB ± 0%    52.5kB ± 0%   -4.52%  (p=0.000 n=10+9)

name                       old allocs/op  new allocs/op  delta
DeriveSha200/std_trie-6       2.87k ± 0%     2.67k ± 0%   -6.97%  (p=0.000 n=10+10)
DeriveSha200/stack_trie-6     1.27k ± 0%     1.07k ± 0%  -15.75%  (p=0.000 n=10+10)
```

### CODE REVIEW

File: rlp/encode.go [hunk 1]
```diff
@@ -604,6 +604,19 @@ func makeEncoderWriter(typ reflect.Type) writer {
 	return w
 }
 
+func PutInt(b []byte, i int) (size int) {
```
Comment (fjl) [inline rlp/encode.go:607]: Please document this.

**Review by fjl — CHANGES_REQUESTED** [review]
Checking this again, I think we should have `AppendInt` instead of `PutInt`. What it should do is: given an input slice, append the RLP integer encoding to it and return the resulting slice. At the call site, you can then simply declare a `[]byte` variable, then call `index = rlp.AppendInt(index[:0], i)`. Subsequent calls will reuse the backing array of `index`.

**Comment by holiman** [comment]
Yes, you are totally right. I had the same thought a few days ago, will update it now

**Comment by holiman** [comment]
Fixed. The latest changes (compared to the first changes):
```
name                       old time/op    new time/op    delta
DeriveSha200/std_trie-6       546µs ± 4%     520µs ± 4%  -4.74%  (p=0.000 n=9+9)
DeriveSha200/stack_trie-6     458µs ± 5%     442µs ± 1%  -3.41%  (p=0.000 n=9+8)

name                       old alloc/op   new alloc/op   delta
DeriveSha200/std_trie-6       274kB ± 0%     274kB ± 0%  -0.02%  (p=0.001 n=10+10)
DeriveSha200/stack_trie-6    52.5kB ± 0%    52.5kB ± 0%    ~     (p=0.067 n=9+10)

name                       old allocs/op  new allocs/op  delta
DeriveSha200/std_trie-6       2.67k ± 0%     2.67k ± 0%    ~     (all equal)
DeriveSha200/stack_trie-6     1.07k ± 0%     1.07k ± 0%    ~     (all equal)

```
And compared to `master`: 
```
name                       old time/op    new time/op    delta
DeriveSha200/std_trie-6       590µs ± 8%     520µs ± 4%  -11.82%  (p=0.000 n=9+9)
DeriveSha200/stack_trie-6     492µs ± 2%     442µs ± 1%  -10.22%  (p=0.000 n=9+8)

name                       old alloc/op   new alloc/op   delta
DeriveSha200/std_trie-6       275kB ± 0%     274kB ± 0%   -0.63%  (p=0.000 n=10+10)
DeriveSha200/stack_trie-6    55.0kB ± 0%    52.5kB ± 0%   -4.53%  (p=0.000 n=10+10)

name                       old allocs/op  new allocs/op  delta
DeriveSha200/std_trie-6       2.87k ± 0%     2.67k ± 0%   -6.97%  (p=0.000 n=10+10)
DeriveSha200/stack_trie-6     1.27k ± 0%     1.07k ± 0%  -15.75%  (p=0.000 n=10+10)
```

**Comment by holiman** [comment]
ping @fjl

**Review by fjl — APPROVED** [review]

**Comment by fjl** [comment]
BTW, I think an additional optimization might be possible: right now, we are encoding the elements to a new byte slice (in `GetRlp`). They could also be encoded into a reusable `bytes.Buffer`. It would need some changes to the DerivableList interface though.

**Comment by holiman** [comment]
I don't think that would work well, the trie needs exclusive access to the elements anyway. Right now, it just assumes it has, but if we reuse buffers for value, we'd have to copy on TryUpdate, and thus the optimization would be mostly moot(?).

**Pull request merged**